README.md
plugin.video.animefrost
=======================

[![Build Status](https://travis-ci.org/nabilbendafi/plugin.video.animefrost.png?branch=master)](https://travis-ci.org/nabilbendafi/plugin.video.animefrost)

Kodi Video Add-on for animefrost.net

This plugin is not related in any way to animefrost.net team or staff member

Disclamer:
----------
All videos of animefrost.net are not hosted on the server that runs this site. Videos are hosted on 3rd party video streaming sites such as but are not limited to: Facebook, DailyMotion, Youtube, Stage6, MyspaceTV, Megavideo and/or Veoh. If you own any work on this site and would like to have them removed, please contact AnimeSeed.com administrators, as well as the 3rd party site which hosts the files. All we are doing is making use of available legal content found all over the internet in a respectful manner. If you have any other issues, please contact us and we’ll try to assist you in the best way possible.


changelog.txt
- CI tests

v0.0.3
- Pydev

v0.0.2
- Fix Google video URL resolution
- Show all video list

v0.0.1
- Initial commit


tests/test_api.py
import random
import unittest
import urlparse

from mechanize import Browser, HTTPError

import animefrost

MAIN_URL = 'http://www.animefrost.net'
USER_AGENT = 'Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.9.0.1) Gecko/2008071615 Fedora/3.0.1-1.fc9 Firefox/3.0.1'

class AnimeFrostTestCase(unittest.TestCase):

    """Test case for Anime Frost api."""

    def setUp(self):
        self.api = animefrost.API()

    def test_instance(self):
        """Tests API instance type."""
        self.assertTrue(isinstance(self.api, animefrost.API))

    def test_browser(self):
        """Tests API's browser."""
        browser = self.api.browser

        self.assertTrue(isinstance(browser, Browser))
        for header in browser.addheaders:
            if header[0] != 'User-agent':
                continue
            print(header)
            self.assertEqual(header[1], USER_AGENT)

        url = MAIN_URL + '/no_such_url'
        self.assertRaises(HTTPError, browser.open, url)

    def test_all(self):
        """Tests all animes list."""
        all_animes = self.api.get_all(7)
        self.assertTrue(all_animes)

        no_animes = self.api.get_all(99)
        self.assertFalse(no_animes)

    def test_categories(self):
        """Tests category list."""
        categories = self.api.get_categories()

        self.assertIn('Martial Arts', [category['label'] for category in categories])

        for category in categories:
            self.assertNotIn('no_such_category', category['label'])

    def test_category(self):
        """Tests anime list."""
        categories = ['action', 'adventure', 'comedy', 'demons', 'drama', 'ecchi',
                      'fantasy', 'game', 'harem', 'historical', 'horror', 'josei',
                      'kids', 'madhouse', 'magic', 'martial-arts', 'mecha',
                      'military', 'music-2', 'mystery', 'parody', 'police',
                      'psychological-2', 'romance', 'samurai', 'school']

        category = random.choice(categories)

        animes = self.api.get_category(category=category)
        self.assertTrue(animes)

        animes = self.api.get_category(page=99, category=category)
        self.assertFalse(animes)

        animes = self.api.get_category(category='not_a_category')
        self.assertFalse(animes)

    def test_animes(self):
        pass

    def test_anime(self):
        """Tests episodes list."""
        animes = ['manyuu-hikenchou', 'bleach']

        anime = random.choice(animes)

        episodes = self.api.get_anime(anime)
        self.assertTrue(episodes)

        episodes = self.api.get_anime(anime='not_an_anime')
        self.assertFalse(episodes)

    def test_episode(self):
        """Tests video url."""
        anime = 'bleach'
        episodes = range(1, 367)

        episode = random.choice(episodes)

        video_url = self.api.get_episode(anime, str(episode))
        self.assertTrue(video_url)

        video_url = self.api.get_episode(anime='not_an_anime', episode=str(episode))
        self.assertFalse(video_url)

        video_url = self.api.get_episode(anime)
        query = urlparse.urlparse(video_url).query
        query_attrs = urlparse.parse_qs(query)
        query_attrs.pop('signature')

        another_video_url = self.api.get_episode(anime, episode='999')
        another_query = urlparse.urlparse(video_url).query
        another_query_attrs = urlparse.parse_qs(another_query)
        another_query_attrs.pop('signature')

        self.assertEqual(query_attrs, another_query_attrs)

    def test_search(self):
        """Tests search result."""
        pattern = 'naruto'

        results = self.api.search(pattern=pattern)

        self.assertTrue(results)
        for result in results:
            self.assertIn(pattern, result['path'])

        results = self.api.search(page=99, pattern=pattern)
        self.assertFalse(results)

        results = self.api.search(pattern='not_an_anime')
        self.assertFalse(results)

    def test_featured(self):
        """Tests featured anime list."""
        featured = self.api.get_featured()

        self.assertTrue(featured)

    def test_latest(self):
        """Tests latest anime list."""
        latest = self.api.get_latest()

        self.assertTrue(latest)

    def tearDown(self):
        pass

if __name__ == '__main__':
    #Rock'n'Roll
    unittest.main()


addon.py
import os
import sys

import xbmcswift2
from xbmcswift2 import xbmcgui as xbmcgui
from fileinput import filename

__addon_name__ = 'Anime Frost'
__addon_id__ = 'plugin.video.animefrost'

plugin = xbmcswift2.Plugin(__addon_name__, __addon_id__, __file__)
addon = plugin.addon

if xbmcswift2.CLI_MODE:
    __addon_path__ = os.getcwd()
else:
    __addon_path__ = addon.getAddonInfo('path')

# append lib directory
sys.path.append(os.path.join(__addon_path__, 'resources', 'lib'))
from animefrost import API as animefrost
api = animefrost()


@plugin.route('/')
def index():
    """Display plugin main menu items."""
    entries = {'A-Z': 'get_all',
               'Featured Videos': 'get_featured',
               'Discover New Anime': 'get_latest',
               'Categories': 'get_categories',
               'Search': 'video_search',
               'FAQ': 'show_faq'}

    items = [{'label': entry,
              'path': plugin.url_for(entries[entry]),
              'is_playable': False
              } for entry in entries]

    return plugin.finish(items)


@plugin.route('/all/')
def get_all():
    """Display the list of all available animes."""
    items = api.get_all()

    items = [{'label': item['label'],
              'path': plugin.url_for('get_anime',
                                     anime=item['path']),
              'thumbnail': item['thumbnail'],
              'is_playable': False
              } for item in items]

    return items


@plugin.cached_route('/categories/')
def get_categories():
    """Display the list of categories."""
    items = api.get_categories()
    print(items)

    items = [{'label': item['label'],
              'path': plugin.url_for('get_category',
                                     category=item['path']),
              'is_playable': False
              } for item in items]

    return items


@plugin.cached_route('/category/<category>')
def get_category(category):
    """
    Display the list of anime for the requested category.

    :param category: category name.
    """
    items = api.get_category(page=1, category=category)

    items = [{'label': item['label'],
              'path': plugin.url_for('get_anime',
                                     anime=item['path']),
              'thumbnail': item['thumbnail'],
              'is_playable': False
              } for item in items]

    return items


@plugin.cached_route('/anime/<anime>')
def get_anime(anime):
    """
    Display the list of episodes for the requested anime.

    :param anime: anime name.
    """
    items = api.get_anime(anime=anime)

    items = [{'label': item['label'],
              'path': plugin.url_for('get_episode',
                                     anime=anime,
                                     episode=item['episode']),
              'is_playable': False
              } for index, item in enumerate(items)]

    return items


@plugin.cached_route('/anime/<anime>/episode/<episode>')
def get_episode(anime, episode='1'):
    """Display the video for the requested anime and episode.

    :param anime:   anime name.
    :param episode: episode number.
    """
    episode = api.get_episode(anime, episode)

    item = {'label': anime,
            'path': plugin.url_for('play',
                                   url=episode),
            'is_playable': True}

    return [item]


@plugin.route('/search/')
def video_search():
    """Display the provided video."""

    search_string = plugin.keyboard()
    if search_string:
        plugin.log.debug('video_search gots a string: "%s"' % search_string)
        url = plugin.url_for('video_search_result',
                             search_string=search_string)
        # immediatly call video_search_result function
        plugin.redirect(url)


@plugin.route('/search/<search_string>/')
def video_search_result(search_string):
    """Display the requested anime.

    :param search_string: anime to look for.
    """
    items = api.search(pattern=search_string)
    print(items)

    if not items:
        msg = 'Sorry, no match found for: ' % search_string
        plugin.log.debug(msg)
        dialog = Dialog()
        dialog.notification(__addon_name__, msg)

    items = [{'label': item['label'],
              'path': plugin.url_for('get_anime',
                                     anime=item['path']),
              'thumbnail': item['thumbnail']
              } for item in items]

    return items


@plugin.cached_route('/featured/')
def get_featured():
    items = api.get_featured()

    items = [{'label': item['label'],
              'path': plugin.url_for('get_anime',
                                     anime=item['path']),
              'thumbnail': item['thumbnail'],
              'is_playable': False
              } for item in items]

    return items


@plugin.cached_route('/latest/')
def get_latest():
    items = api.get_latest()

    items = [{'label': item['label'],
              'path': plugin.url_for('get_anime',
                                     anime=item['path']),
              'thumbnail': item['thumbnail'],
              'is_playable': False
              } for item in items]

    return items


@plugin.route('/play/<url>')
def play(url):
    """Play the video.

    :param url: playable video URL
    """
    plugin.log.info('Playing url: %s' % url)
    return plugin.set_resolved_url(url)


@plugin.route('/faq/')
def show_faq():
    """Simply show a popup window, with some information."""
    dialog = xbmcgui.Dialog()
    filename = os.path.join(__addon_path__, 'faq')
    try:
        faq = open(filename)
        dialog.ok("FAQ", faq.read())
    except IOError:
        pass
    finally:
        faq.close()


if __name__ == '__main__':
    try:
        remote_debugger = addon.getSetting('remote_debugger')
        # default false
        remote_debugger_host = addon.getSetting('remote_debugger_host')
        # default '127.0.0.1'
        remote_debugger_port = addon.getSetting('remote_debugger_port')
        # default 5678

        # append pydev remote debugger
        if remote_debugger == 'true':
            import pydevd
            # stdoutToServer and stderrToServer redirect stdout and stderr
            # to eclipse console
            pydevd.settrace(remote_debugger_host, port=remote_debugger_port,
                            stdoutToServer=True, stderrToServer=True)
    except ImportError:
        plugin.log.error("Error: make sure pydevd is installed !")
        sys.exit(1)
    except:
        pass

    # Rock'n'Roll
    try:
        plugin.run()
    except Exception, e:
        plugin.notify(msg=e)


resources/lib/animefrost.py
import inspect
import re
import os
import urllib
import urlparse

import sys
reload(sys)
sys.setdefaultencoding("utf-8")

from BeautifulSoup import BeautifulSoup
from mechanize import Browser

MAIN_URL = 'http://www.animefrost.net'
USER_AGENT = 'Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.9.0.1) Gecko/2008071615 Fedora/3.0.1-1.fc9 Firefox/3.0.1'


class API():

    def __init__(self):
        """Constructor."""
        self.browser = Browser()
        """browser"""

        # disable robot.txt usage
        self.browser.set_handle_robots(False)
        self.browser.addheaders = [('User-agent', USER_AGENT)]

    def get_all(self, page=1):
        """Find all anime available.

        :returns: list of animes
        """
        url = 'alphabetical-image-list/page/%(page)s' % {'page': page}

        main_elem = self.get_html_tree(url)

        return self.get_animes(main_elem, page)

    def get_category(self, page=1, category=None):
        """Find all anime available from a specific category.

        :returns: list of animes
        """
        url = 'video_tag/%(category)s/page/%(page)d' % {'category': category,
                                                        'page': page}

        main_elem = self.get_html_tree(url)

        return self.get_animes(main_elem, page, category)

    def get_categories(self):
        """Find all categories available.

        :returns: list of categories
        """
        categories = []

        tags = ['Action', 'Adventure', 'Comedy', 'Demons', 'Drama',
                'Ecchi', 'Fantasy', 'Game', 'Harem', 'Historical',
                'Horror', 'Josei', 'Kids', 'Madhouse', 'Magic',
                'Martial Arts', 'Mecha', 'Military', 'Music', 'Mystery',
                'Parody', 'Police', 'Psychological', 'Romance', 'Samurai',
                'School']

        urls = ['action', 'adventure', 'comedy', 'demons', 'drama', 'ecchi',
                'fantasy', 'game', 'harem', 'historical', 'horror', 'josei',
                'kids', 'madhouse', 'magic', 'martial-arts', 'mecha',
                'military', 'music-2', 'mystery', 'parody', 'police',
                'psychological-2', 'romance', 'samurai', 'school']

        for index, tag in enumerate(tags):
            category = {'label': tag,
                        'path': urls[index]}
            categories.append(category)

        return categories

    def get_animes(self, html_elem, *args):
        animes = []
        last_page = 1

        if not html_elem:
            return animes

        # find last page index
        last_page_elem = html_elem.findAll('a', {'class': 'page-numbers'})
        if last_page_elem:
            last_page = int(last_page_elem[-1].getText())

        div_elems = html_elem.findAll('div', {'class': lambda x: x
                                                       and 'item' in x.split()})
        for div_elem in div_elems:
            a_elem = div_elem.findAll('a')[-1]
            img_elem = div_elem.find('img')

            # <scheme>://<netloc>/<path>?<query>#<fragment>
            path = urlparse.urlsplit(a_elem.get('href'))[2]
            regex = re.compile('/watch/(?P<anime>.*)/')

            anime = {'label': a_elem.getText(),
                     'path': re.search(regex, path).groupdict()['anime'],
                     'thumbnail': img_elem.get('src')}
            animes.append(anime)

        # always first arg of caller function
        args = list(args)
        page = args[0]

        # recursive call for other pages
        if page < last_page:
            stack = inspect.stack()
            parent_func = stack[1][3]
            # call caller function
            args[0] = page + 1
            # with original args (increment page number)
            animes.extend(getattr(self, parent_func)(*args))

        return animes

    def search(self, page=1, pattern=''):
        pattern = urllib.quote(pattern)

        url = 'page%(page)d?s=%(pattern)s' % {'pattern': pattern,
                                              'page': page}

        search = []

        search_elem = self.get_html_tree(url)

        if not search_elem:
            return search

        div_elem = search_elem.find('div', {'class': 'alert alert-info'})
        # no anime found matching search
        if div_elem:
            return search
        else:
            search = self.get_animes(search_elem, page, pattern)

        return search

    def get_anime(self, anime):
        """Display the list of episodes for the requested anime.

        :param anime: anime name

        :returns: list of episodes.
        """
        episodes = []

        url = 'watch/%(anime)s' % {'anime': anime}

        main_elem = self.get_html_tree(url)

        if not main_elem:
            return episodes

        ul_elem = main_elem.find('ul', {'class': 'pagination post-episode'})
        a_elems = ul_elem.findAll('a')
        for a_elem in a_elems:
            episode_num = a_elem.text
            episode = {'label': 'Episode %d' % int(episode_num),
                       'episode': episode_num}
            episodes.append(episode)

        return episodes

    def get_episode(self, anime, episode='1'):
        """Display the video link for the requested anime and episode.

        :param anime:   anime name
        :param episode: episode number

        :returns: video link.
        """
        url = 'watch/%(anime)s/?episode=%(episode)s' % {'anime': anime,
                                                        'episode': episode}

        video_url = ''

        main_elem = self.get_html_tree(url)

        if not main_elem:
            return video_url

        iframe_elem = main_elem.find('iframe')
        video_url = iframe_elem.get('src')

        # 2-step video url retrieval
        pattern = 'https?://(docs|drive).google.com/file/d/(?P<video_id>.*)/preview'
        regex = re.compile(pattern)
        video_id = re.search(regex, video_url)

        if video_id:
            video_id = video_id.groupdict()['video_id']

            url = 'https://drive.google.com/file/d/%(video_id)s/' % {'video_id': video_id}
            main_elem = self.get_html_tree(url)

            for script_elem in main_elem.findAll('script'):
                text = script_elem.text
                if 'fmt_stream_map' in text:
                    pattern = '\|(?P<link>https:[^|]*)\|https.*'
                    regex = re.compile(pattern)
                    video_url = re.findall(regex, text)[0]

                    # Some coding/decode stuff
                    video_url = video_url.decode('unicode_escape')
                    video_url = video_url.encode('utf-8')
                    break

        return video_url

    def get_featured(self):
        """Find all featured anime available."""
        animes = []

        main_elem = self.get_html_tree()

        if not main_elem:
            return animes

        feature_elem = main_elem.find('div', {'class': 'featured-wrapper'})
        item_elems = feature_elem.findAll('div', {'class': 'item-img'})
        for item_elem in item_elems:
            a_elem = item_elem.find('a')
            img_elem = a_elem.find('img')

            path = urlparse.urlsplit(a_elem.get('href'))[2]
            regex = re.compile('/watch/(?P<anime>.*)/')

            anime = {'label': a_elem.get('title'),
                     'path': re.search(regex, path).groupdict()['anime'],
                     'thumbnail': img_elem.get('src')}
            animes.append(anime)

        return animes

    def get_latest(self):
        """Find all latest anime available."""
        animes = []

        main_elem = self.get_html_tree()

        if not main_elem:
            return animes

        feature_elem = main_elem.find('div', {'class': 'latest-wrapper'})
        item_elems = feature_elem.findAll('div', {'class': 'item-img'})
        for item_elem in item_elems:
            a_elem = item_elem.find('a')
            img_elem = a_elem.find('img')

            path = urlparse.urlsplit(a_elem.get('href'))[2]
            regex = re.compile('/watch/(?P<anime>.*)/')

            anime = {'label': a_elem.get('title'),
                     'path': re.search(regex, path).groupdict()['anime'],
                     'thumbnail': img_elem.get('src')}
            animes.append(anime)

        return animes

    def get_html_tree(self, endpoint=''):
        """Return HTML tree as url is browsed.

        :param endpoint: URL endpoint to browse

        :returns: an HTML tree
        """
        html = ""
        try:
            url = urllib.basejoin(MAIN_URL, endpoint)
            self.browser.open(url)
            # retrieve page content
            html = self.browser.response().read()
        except Exception, e:
            pass
        else:
            # soup-it
            html_tree = BeautifulSoup(
                html, convertEntities=BeautifulSoup.HTML_ENTITIES)
            return html_tree


if __name__ == '__main__':
    # Rock'n'Roll
    api = API()
